File: solutions/javascript/vehicle-purchase/2/vehicle-purchase.js
// @ts-check

/**
 * Determines whether or not you need a license to operate a certain kind of vehicle.
 *
 * @param {string} kind
 * @returns {boolean} whether a license is required
 */
export const needsLicense = (kind) => ["car", "truck"].includes(kind);

/**
 * Helps choosing between two options by recommending the one that
 * comes first in dictionary order.
 *
 * @param {string} option1
 * @param {string} option2
 * @returns {string} a sentence of advice which option to choose
 */
export const chooseVehicle = (option1, option2) => {
  const choice = option1.localeCompare(option2) < 1 ? option1 : option2;

  return `${choice} is clearly the better choice.`;
};

/**
 * Calculates an estimate for the price of a used vehicle in the dealership
 * based on the original price and the age of the vehicle.
 *
 * @param {number} originalPrice
 * @param {number} age
 * @returns {number} expected resell price in the dealership
 */
export const calculateResellPrice = (originalPrice, age) => {
  const priceFactor =
    new Map([
      [age < 3, 0.8],
      [age > 10, 0.5],
    ]).get(true) ?? 0.7;

  return originalPrice * priceFactor;
};

File: solutions/cpp/vehicle-purchase/1/vehicle_purchase.cpp
#include "vehicle_purchase.h"

namespace vehicle_purchase
{

    // needs_license determines whether a license is needed to drive a type of
    // vehicle. Only "car" and "truck" require a license.
    bool needs_license(std::string kind)
    {
        return kind == "car" || kind == "truck";
    }

    // choose_vehicle recommends a vehicle for selection. It always recommends the
    // vehicle that comes first in lexicographical order.
    std::string choose_vehicle(std::string a, std::string b)
    {
        std::string choice = a < b ? a : b;

        return choice +
               " is clearly the better choice.";
    }

    // calculate_resell_price calculates how much a vehicle can resell for at a
    // certain age.
    double calculate_resell_price(double original_price, double age)
    {
        double percent = age < 3 ? 0.8 : age >= 10 ? 0.5
                                                   : 0.7;

        return original_price * percent;
    }

} // namespace vehicle_purchase
